pkg/admissioncontroller/webhooks/admission/auditpolicy/admission.go

```diff
@@ -0,0 +1,294 @@
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+
+		if hasAuditPolicy(oldShoot) && reflect.DeepEqual(cmRef, oldShoot.Spec.Kubernetes.KubeAPIServer.AuditConfig.AuditPolicy.ConfigMapRef) {
+			return acadmission.Allowed("audit policy reference did not change in Shoot resource")
+		}
+	}
+
+	auditPolicyCm := &corev1.ConfigMap{}
+	if err := h.apiReader.Get(ctx, kutil.Key(shoot.Namespace, cmRef.Name), auditPolicyCm); err != nil {
+		if apierrors.IsNotFound(err) {
+			return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("referenced audit policy does not exist: namespace: %s, name: %s", cmRef.Namespace, cmRef.Name))
+		}
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("could not retrieve config map: %s", err))
+	}
+	auditPolicy, err := getAuditPolicy(auditPolicyCm)
+	if err != nil {
+		return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("Error: %s in %v/%v", err, cmRef.Namespace, cmRef.Name))
+	}
+
+	// Validate audit policy semantically
+	auditPolicyObj, schemaVersion, err := decoder.Decode([]byte(auditPolicy), nil, nil)
+	if err != nil {
+		return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("failed to decode the provided audit policy err=%v", err))
+	}
+	auditPolicyInternal, ok := auditPolicyObj.(*audit_internal.Policy)
+	if !ok {
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failure to cast to audit Policy type: %v", schemaVersion))
+	}
+	errList := auditvalidation.ValidatePolicy(auditPolicyInternal)
+	if len(errList) != 0 {
+		return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("provided invalid audit policy err=%v", errList))
+	}
+
+	// Validate audit policy schema version against k8s version of shoot
+	if isValidVersion, err := IsValidAuditPolicyVersion(shoot.Spec.Kubernetes.Version, schemaVersion); err != nil {
+		return admission.Errored(http.StatusUnprocessableEntity, err)
+	} else if !isValidVersion {
+		err := fmt.Errorf("your shoot cluster version %q is not compatible with audit policy version %q", shoot.Spec.Kubernetes.Version, schemaVersion.GroupVersion().String())
+		return admission.Errored(http.StatusUnprocessableEntity, err)
+	}
+
+	return acadmission.Allowed("referenced audit policy is valid")
+}
+
+func (h *handler) admitConfigMap(ctx context.Context, request admission.Request) admission.Response {
+	var (
+		oldCm = &corev1.ConfigMap{}
+		cm    = &corev1.ConfigMap{}
+	)
+	if err := h.getOldObject(request, oldCm); err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	if request.Operation == admissionv1.Delete {
+		if controllerutil.ContainsFinalizer(oldCm, shoot.FinalizerName) {
+			return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("deletion of audit-policy is forbidden as it is still referenced by a shoot (cm has finalizer)"))
+		}
+		return acadmission.Allowed("deletion of audit-policy accepted")
+	}
```

Comment:
Hm, do we need this? A finalizer is already set and it will prevent the resource to be removed from the store as long as it is used by any shoot.

Reply:
Yeah, I just wanted to write a similar comment. Is there a reason why you added this?

Reply:
The only reason that come to my mind is if the configmap become immutable after the deletionTimestamp is set, but this is not the case.

Reply:
I agree with you.